src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTests.scala

```diff
@@ -407,6 +409,79 @@ class E2EHyperspaceRulesTests extends HyperspaceSuite with SQLHelper {
       sortedRowsWithHyperspaceDisabled.sameElements(getSortedRows(dfAfterHyperspaceDisabled)))
   }
 
+  test("Verify Join Index Rule utilizes indexes correctly after incremental refresh.") {
+    withTempDir { testDir =>
+      val refreshTestLocation = testDir + "/refresh"
+      withSQLConf(IndexConstants.REFRESH_APPEND_ENABLED -> "true") {
+        // Setup. Create Data.
+        val indexConfig = IndexConfig(s"index", Seq("c2"), Seq("c4"))
+        import spark.implicits._
+        SampleData.testData
+          .toDF("c1", "c2", "c3", "c4", "c5")
+          .limit(10)
+          .write
+          .parquet(refreshTestLocation)
+        val df = spark.read.load(refreshTestLocation)
+
+        // Create Index.
+        hyperspace.createIndex(df, indexConfig)
+
+        // Append to Original Data.
+        SampleData.testData
+          .toDF("c1", "c2", "c3", "c4", "c5")
+          .limit(3)
+          .write
+          .mode("append")
+          .parquet(refreshTestLocation)
+
+        // Refresh Index.
+        hyperspace.refreshIndex(indexConfig.indexName)
+
+        // Create a Join Query.
+        val leftDf = spark.read.parquet(refreshTestLocation)
+        val rightDf = spark.read.parquet(refreshTestLocation)
+
+        def query(): DataFrame = {
+          leftDf
+            .join(rightDf, leftDf("c2") === rightDf("c2"))
+            .select(leftDf("c2"), rightDf("c4"))
+        }
+
+        // Verify Indexes are used, and all index files are picked.
+        verifyIndexUsage(
+          query,
+          getIndexFilesPath(indexConfig.indexName, Seq(0, 1)) ++
+            getIndexFilesPath(indexConfig.indexName, Seq(0, 1)))
+
+        // Verify Bucketing works as expected. This is reflected in shuffle nodes being eliminated
+        // when hyperspace is enabled. Also verify that sort nodes are not removed.
+        // Without Hyperspace.
+        spark.disableHyperspace()
+        val dfWithHyperspaceDisabled = query()
+        var shuffleNodes = dfWithHyperspaceDisabled.queryExecution.executedPlan.collect {
+          case s: ShuffleExchangeExec => s
+        }
+        assert(shuffleNodes.size == 2)
+        var sortNodes = dfWithHyperspaceDisabled.queryExecution.executedPlan.collect {
+          case s: SortExec => s
+        }
+        assert(sortNodes.size == 2)
+
+        // With Hyperspace.
+        spark.enableHyperspace()
+        val dfWithHyperspaceEnabled = query()
+        shuffleNodes = dfWithHyperspaceEnabled.queryExecution.executedPlan.collect {
+          case s: ShuffleExchangeExec => s
+        }
+        assert(shuffleNodes.isEmpty)
+        sortNodes = dfWithHyperspaceEnabled.queryExecution.executedPlan.collect {
```

Comment:
`// SortExec is expected to be present because there are multiple files per bucket.`

Reply:
thanks @imback82 , fixed